---
title: "Gene Expression Prediction Challenge 3.0 (expred3.0)"
subtitle: "sujet et compte rendu"
author: "Florent Chuffart - Team 0"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
``` 

# Objectif

1. On cherche à expliquer l’expression du gène ALS2 dans le jeu de données data_train,
2. Pour rédire les valeurs d’expression du gène ALS2 dans le jeu de données data_test.

# Statistiques descriptives

**Le jeu de donnée `data_train`**

```{r loading_data, echo=TRUE, results="verbatim"}
data_train <- readRDS(file = "data_train.rds")
data_test <- readRDS(file = "data_test.rds")
head(data_train[,1:6])
```

**Distribution de l’expression de ALS2**

```{r}
plot(density(data_train$ALS2))
```

**Distribution du transcriptome dans `data_train`**

```{r}
plot(density(as.matrix(data_train[,4:1003])))
```

**Distribution du methylome dans `data_train`**

Pas utilisé dans ce TP.

```{r}
plot(density(as.matrix(data_train[,1004:2003])))
```

# Method 1 : SIS

La method SIS [Shurely Independant Screening, Zhang HH. J R Stat Soc Series B Stat Methodol. 2008] appliquée au *transcriptome* (définir) consiste à réaliser autant de regressions linéaires simples qu’il y a de variables explicatives (gènes).

*Critères*, pour chaque gène $g$ on calcule le $\beta_{g}$ et la p-valeur $pv_g$ associée. 

*Selection de variables*, on selectionne sur le *volcano plot* les gènes qui ont les meilleurs scores, indépendamment les uns des autres.

*Modèle*, on propose un modèle de regression linéaire multiple expliquant l’expression de ALS2.

**Calcul des modèles indépendants**

Pour chaque gènes calculer le $\beta_{gene}$ et la p-valeur associée au modèle $ALS2~gène$, on charactérisera aussi la distribution des résidus.

```{r sis_compute_models, echo=TRUE}
gs = colnames(data_train)[5:1003]
gs
# Go!
for (i in 1:length(gs)) {
  g = gs[i]
  print(g)
  m = lm(data_train[,"ALS2"]~data_train[,g])
}


pval_shapiro = c()
pval_fisher = c()
beta = c()
for (g in gs) {
  print(g)
  m = lm(data_train[,"ALS2"]~data_train[,g])
  res = m$residuals
  pval_shapiro = c(pval_shapiro,shapiro.test(res)$p.value)
  pval_fisher = c(pval_fisher, anova(m)[1,5])
  beta = c(beta, m$coefficients[[2]])
}
names(pval_shapiro) = gs
names(pval_fisher)  = gs
names(beta)         = gs


```

**Normalité**

Afficher : 

- en abscisse les $-log10(p-valeurs)$ des tests de Shapiro-Wilk sur les résidus de chacun des modèles indépendants,
- en ordonée les $-log10(p-valeurs)$ des tests de Student (ou Fisher) associés.

```{r sis_normalité}
# Go!
plot(-log10(pval_shapiro), -log10(pval_fisher))

```
Commenter.


**Volcano plot**

Afficher : 

- en abscisse les $\beta_{gene}$ des modéles de régression simple pour chaque gène $g$,
- en ordonée les $-log10(p-valeurs)$ des tests de Student (ou Fisher) associés.

```{r sis_compute_volcano}
# Go!
layout(1, respect=TRUE)
plot(beta, -log10(pval_fisher), col="grey")
text(beta, -log10(pval_fisher), names(beta))

g = "RGS7BP"
m = lm(data_train[,"ALS2"]~data_train[,g])
plot(data_train[,g], data_train[,"ALS2"], main=g)
abline(m, col=2)
summary(m)$r.squared

m = lm(ALS2~RGS7BP,data_train)

plot(data_train[,"MFAP3"], data_train[,"RGS7BP"], main=g)

plot(data_train[,"VMO1"], data_train[,"RGS7BP"], main=g)




pairs(data_train[,names(sort(pval_fisher)[1:8])])

```
Commenter.



**Sélection de variable

Enumérez vos critères.

Ecrire la formulation analystique de votre modèles.



```{r echo=TRUE, results=verbatim}
selected_genes = names(sort(pval_fisher)[1:8])
selected_genes

m1 = lm(ALS2~RGS7BP,data_train)
summary(m1)$r.squared

m8 = lm(ALS2~RGS7BP+MFAP3+PRRC2C+VMO1+SCRN3+NRIP1+PCNX1+TMUB1, data_train)
summary(m8)$r.squared

selected_genes = names(sort(pval_fisher)[1:12])
formula = as.formula(paste0("ALS2~", paste0(selected_genes, collapse="+")))
formula
m12 = lm(formula, data_train)
summary(m12)$r.squared


r2 = c()
for (i in 1:12) {
  selected_genes = names(sort(pval_fisher)[1:i])
  formula = as.formula(paste0("ALS2~", paste0(selected_genes, collapse="+")))
  formula
  mi = lm(formula, data_train)
  r2 = c(r2,summary(mi)$r.squared)
}

plot(r2)
```

Tracer :
  - en abscisse i de 1 à 12
  - en ordonnée `summary(mi)$r.squared`


**Prédiction**

1. Prédisez les valeurs de ALS2 du jeu de donnée `data_train`
2. Prédisez les valeurs de ALS2 du jeu de donnée `data_test`
3. Score obtenu sur codalab
4. Comparer

```{r sis_pred}
"RMSE" = function(data_truth, data_pred) {
    # Root Mean Square Error
    return(sqrt(mean((data_truth - data_pred)^2)))
}
# Go!
data_truth = data_train$ALS2
# data_pred = predict(data_train...)
# score =  RMSE(data_truth, data_train)

m0 = lm(ALS2~1, data_train)
data_pred = predict(m0, data_train, type="response")
data_truth = data_train$ALS2
err0_train0 = RMSE(data_truth, data_pred)
err0_train0
err0_test0

m1 = lm(ALS2~RGS7BP, data_train)
data_pred = predict(m1, data_train, type="response")
data_truth = data_train$ALS2
err0_train1 = RMSE(data_truth, data_pred)
err0_train1
err0_test1

m2 = lm(ALS2~RGS7BP+MFAP3, data_train)
data_pred = predict(m2, data_train, type="response")
data_truth = data_train$ALS2
err0_train2 = RMSE(data_truth, data_pred)
err0_train2
err0_test2

err_train = c()
for (i in 1:8) {
  selected_genes = names(sort(pval_fisher)[1:i])
  formula = as.formula(paste0("ALS2~", paste0(selected_genes, collapse="+")))
  formula
  mi = lm(formula, data_train)
  data_pred = predict(mi, data_train, type="response")
  data_truth = data_train$ALS2
  err_train[i] = RMSE(data_truth, data_pred)
}
 
err_test = c()
err_test[1] = 0.2800726811
err_test[2] = 0.2370102069
err_test[3] = 0.2351253487
err_test[4] = 0.2299540168
err_test[5] = 0.2142468642
err_test[6] = 0.2169962834
err_test[7] = 0.2159720152
err_test[8] = 0.2154245011  




plot( 1:8, err_test, ylim=c(0,0.4), col=2, type="l")
lines(1:8, err_train, col=4)
legend("bottomleft", col=c(2,4), c("test", "train"), lty=1)
....



  


```

**Commentaires



# Method 2 : ACP


```{r pca2, eval=TRUE}
d = as.matrix(data_train[,1004:2003])
pca = prcomp(d, scale=TRUE)

v = pca$sdev * pca$sdev
p = v / sum(v) * 100

layout(matrix(1:6,2), respect=TRUE)
# layout(matrix(1:2,1), respect=TRUE)
barplot(p)

# data_train$histo2 = as.factor(substr(data_train$histology, 1, 5))
data_train$histo2 = data_train$histology
for (i in 1:5) {
  j = i+1
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), pch=16, col=as.numeric(data_train[rownames(pca$x),]$histo2))
  scale_factor = min(abs(c(min(c(pca$x[,i], pca$x[,j])), max(c(pca$x[,i], pca$x[,j])))))  
  # scale_factor = min(abs(c(max(min(pca$x[,i]), min(pca$x[,j])), min(max(pca$x[,i]), max(pca$x[,j])))))
  plotrix::draw.ellipse(0,0,scale_factor,scale_factor, lty=2, border="grey")
  # arrows(0,0,pca$rotation[,i]*scale_factor, pca$rotation[,j]*scale_factor, col="grey")
  # text(pca$rotation[,i]*scale_factor, pca$rotation[,j]*scale_factor, rownames(pca$rotation))
}
```

```{r pairs, fig.height=9}
pairs(pca$x[,1:8], pch=".", col=as.numeric(data_train[rownames(pca$x),]$histo2))
```

```{r hm, eval=FALSE}
data = t(as.matrix(data_train[,0004:1003]))
source("~/projects/epimedpipelines/results/commons.R")
plot_expr_hm(data=data, normalization="zscore_rows",   colors=c("cyan", "cyan", "black", "red", "red"))
```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```



